### PR TITLE
Allow txn relay in need blocks state for PoS

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
* **#1160** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1160?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `feature/proof-of-stake`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @lazynina and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/server.go [hunk 1]
```diff
@@ -1930,12 +1930,19 @@ func (srv *Server) _addNewTxn(
 		return nil, err
 	}
 
-	if srv.blockchain.chainState() != SyncStateFullyCurrent {
+	srv.blockchain.ChainLock.RLock()
+	tipHeight := uint64(srv.blockchain.BlockTip().Height)
```
Comment (tholonious) [inline lib/server.go:1934]: Cache the chain state so it doesn't shift from underneath you
```suggestion
	tipHeight := uint64(srv.blockchain.BlockTip().Height)
	chainState := srv.blockchain.chainState()
```

File: lib/server.go [hunk 2]
```diff
@@ -1930,12 +1930,19 @@ func (srv *Server) _addNewTxn(
...
+	srv.blockchain.ChainLock.RUnlock()
 
-		err := fmt.Errorf("Server._addNewTxnAndRelay: Cannot process txn "+
-			"from peer %v while syncing: %v %v", pp, srv.blockchain.chainState(), txn.Hash())
-		glog.Error(err)
-		return nil, err
+	if srv.blockchain.chainState() != SyncStateFullyCurrent {
```
Comment (tholonious) [inline lib/server.go:1937]:
```suggestion
	if chainState != SyncStateFullyCurrent {
```

File: lib/server.go [hunk 3]
```diff
@@ -1930,12 +1930,19 @@ func (srv *Server) _addNewTxn(
...
+		// We allow txn relay if chain is in a need blocks state and is running PoS
+		if !(srv.blockchain.chainState() == SyncStateNeedBlocksss &&
+			srv.blockchain.params.IsPoSBlockHeight(tipHeight)) {
```
Comment (tholonious) [inline lib/server.go:1940]: The logic seems right but the negation on the whole condition makes it a bit harder to interpret. This should work just as well
```suggestion
		if chainState != SyncStateNeedBlocksss || !srv.blockchain.params.IsPoSBlockHeight(tipHeight)) {
```

File: lib/server.go [hunk 4]
```diff
@@ -1930,12 +1930,19 @@ func (srv *Server) _addNewTxn(
...
+	srv.blockchain.ChainLock.RLock()
+	tipHeight := uint64(srv.blockchain.BlockTip().Height)
+	srv.blockchain.ChainLock.RUnlock()
 
-		err := fmt.Errorf("Server._addNewTxnAndRelay: Cannot process txn "+
-			"from peer %v while syncing: %v %v", pp, srv.blockchain.chainState(), txn.Hash())
-		glog.Error(err)
-		return nil, err
+	if srv.blockchain.chainState() != SyncStateFullyCurrent {
+		// We allow txn relay if chain is in a need blocks state and is running PoS
```
Comment (tholonious) [inline lib/server.go:1938]:
```suggestion
		// We allow txn relay if chain is in a need blocks state and is running PoS. We will error in two
		// cases:
		// - The chainState is not need blocks state
		// - The chainState is the need blocks state, but the chain is not on PoS
```

**Review by tholonious — APPROVED** [review]

**Pull request merged**